src/Craft/Systemd.hs

```diff
@@ -0,0 +1,384 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Craft.Systemd where
+
+import Control.Lens
+import Data.List.Utils (join)
+
+import Craft
+import Craft.File
+
+-- Helpful: https://www.digitalocean.com/community/tutorials/understanding-systemd-units-and-unit-files
+
+-- A service name is simply the name of that service
+type ServiceName = String
+
+-- All systemd Units can be transformed to their plaintext output
+class WritableUnit a where
+  transformUnit :: a -> String
+
+-- A generic implementation for the String type
+instance WritableUnit String where
+  transformUnit sn = show sn
+
+-- A generic implementation for the Bool type
+instance WritableUnit Bool where
+  transformUnit True = "true"
+  transformUnit False = "false"
+
+-- A generic implementation for the Int type
+instance WritableUnit Int where
+  transformUnit i = show i
+
+-- Composite units need to be able to furnish a filename (*without* a path)
+class CompositeUnit unit where
+  getFileName :: unit -> FilePath
+
+-- | Helpers |--
+-- =====================================================================
+-- Functions for extracting optional values to be transformed
+writeOptionalSection :: WritableUnit a => Maybe a -> String
+writeOptionalSection unit =
+  case unit of
+    {
+      Nothing -> "";
+      Just u -> transformUnit u ++ "\n"
+    }
+
+-- Optionally write a directive
+writeOptional :: WritableUnit a => String -> Maybe a -> String
+writeOptional directive optionalParam =
+  case optionalParam of
+    {
+      Nothing -> "";
+      Just param -> directive ++ "=" ++ transformUnit param ++ "\n"
+    }
+
+writeOptionalList :: WritableUnit a => String -> Maybe [a] -> String
+writeOptionalList directive optionalParams =
+  case optionalParams of
+    {
+      Nothing -> "";
+      Just params -> directive ++ "=" ++
+      (join " " (map transformUnit params)) ++ "\n"
+    }
+-- =====================================================================
+
+-- | [Unit] |--
+-- =====================================================================
+data UnitSection = UnitSection { _description   :: String
+                               , _documentation :: Maybe String
+                               , _requires      :: Maybe [ServiceName]
+                               , _wants         :: Maybe [ServiceName]
```

Comment:
> Any ideas how to express an optional string then?

Ugh, yeah, this is one of the worst warts in Haskell, for example `head ""` is a runtime error. I think for `String` it's fine to use `Maybe String` even though it's just a type synonym for `Maybe [Char]`. To avoid problems, we can handle `Nothing` and `Just ""` the same way.